near-sdk/src/store/tree_map/mod.rs

```diff
@@ -0,0 +1,2059 @@
+mod entry;
+mod impls;
+mod iter;
+
+pub use entry::Entry;
+pub use iter::{Iter, IterMut, Keys, Range, RangeMut, Values, ValuesMut};
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::borrow::Borrow;
+use std::ops::RangeBounds;
+
+use super::lookup_map as lm;
+use crate::crypto_hash::{CryptoHasher, Sha256};
+use crate::store::free_list::{FreeList, FreeListIndex};
+use crate::store::LookupMap;
+use crate::{env, IntoStorageKey};
+
+type NodeAndIndex<'a, K> = (FreeListIndex, &'a Node<K>);
+
+fn expect<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| env::abort())
+}
+
+/// TreeMap based on AVL-tree
+///
+/// Runtime complexity (worst case):
+/// - `get`/`contains_key`:     O(1) - LookupMap lookup
+/// - `insert`/`remove`:        O(log(N))
+/// - `min`/`max`:              O(log(N))
+/// - `above`/`below`:          O(log(N))
+/// - `range` of K elements:    O(Klog(N))
+///
+pub struct TreeMap<K, V, H = Sha256>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    values: LookupMap<K, V, H>,
+    tree: Tree<K>,
+}
+
+//? Manual implementations needed only because borsh derive is leaking field types
+// https://github.com/near/borsh-rs/issues/41
+impl<K, V, H> BorshSerialize for TreeMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn serialize<W: borsh::maybestd::io::Write>(
+        &self,
+        writer: &mut W,
+    ) -> Result<(), borsh::maybestd::io::Error> {
+        BorshSerialize::serialize(&self.values, writer)?;
+        BorshSerialize::serialize(&self.tree, writer)?;
+        Ok(())
+    }
+}
+
+impl<K, V, H> BorshDeserialize for TreeMap<K, V, H>
+where
+    K: BorshSerialize + Ord,
+    V: BorshSerialize,
+    H: CryptoHasher<Digest = [u8; 32]>,
+{
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, borsh::maybestd::io::Error> {
+        Ok(Self {
+            values: BorshDeserialize::deserialize(buf)?,
+            tree: BorshDeserialize::deserialize(buf)?,
+        })
+    }
+}
+
+#[derive(BorshDeserialize, BorshSerialize)]
+struct Tree<K>
+where
+    K: BorshSerialize,
+{
+    root: Option<FreeListIndex>,
+    nodes: FreeList<Node<K>>,
+}
+
+impl<K> Tree<K>
+where
+    K: BorshSerialize + Ord,
+{
+    fn new<S>(prefix: S) -> Self
+    where
+        S: IntoStorageKey,
+    {
+        Tree { root: None, nodes: FreeList::new(prefix) }
+    }
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize, Debug)]
+struct Node<K> {
+    key: K,                     // key stored in a node
+    lft: Option<FreeListIndex>, // left link of a node
+    rgt: Option<FreeListIndex>, // right link of a node
+    ht: u32,                    // height of a subtree at a node
```

Comment:
I know you have inherited the AVL implementation from the old collection and this optimization is probably not worth the trouble, but instead of storing tree height, you can actually just store the balance factor (-1, 0, 1). It is quite easy to see that it would be enough by looking at all current usages of `ht`. From what I remember upholding the balance factor invariant is a bit tricky though and the gain is minuscule (if any at all), but I thought I would share my thoughts just in case.

Reply:
let's tackle this in a following PR. This is under `unstable` so let's try to refactor and optimize once we have a framework to compare the optimizations

Reply:
Yeah, I don't mind. Just thought I would point this out before I forget :)